### PR TITLE
Enable the clippy::arithmetic_side_effects lint

### LINKED ISSUE #160: Fix failures for lint clippy::arithmetic_side_effects
# Clippy reports for lint clippy::arithmetic_side_effects

Issues found:
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/ec-core/src/operator/composable/repeat_with.rs#L59C16-L59C25">packages/ec-core/src/operator/composable/repeat_with.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
  --> packages/ec-core/src/operator/composable/repeat_with.rs:59:16
   |
59 |             Ok(input + 1)
   |                ^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
   = note: requested on the command line with `-W clippy::arithmetic-side-effects`
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/ec-core/src/operator/composable/then.rs#L42C16-L42C25">packages/ec-core/src/operator/composable/then.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
  --> packages/ec-core/src/operator/composable/then.rs:42:16
   |
42 |             Ok(input + 1)
   |                ^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/ec-core/src/operator/composable/then.rs#L52C16-L52C25">packages/ec-core/src/operator/composable/then.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
  --> packages/ec-core/src/operator/composable/then.rs:52:16
   |
52 |             Ok(input * 2)
   |                ^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/ec-linear/src/mutator/umad.rs#L124C17-L124C33">packages/ec-linear/src/mutator/umad.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
   --> packages/ec-linear/src/mutator/umad.rs:124:17
    |
124 |                 short_index += 1;
    |                 ^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
    = note: requested on the command line with `-W clippy::arithmetic-side-effects`
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/ec-linear/src/mutator/umad.rs#L126C17-L126C33">packages/ec-linear/src/mutator/umad.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
   --> packages/ec-linear/src/mutator/umad.rs:126:17
    |
126 |                 num_missing += 1;
    |                 ^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/ec-linear/src/mutator/umad.rs#L128C13-L128C28">packages/ec-linear/src/mutator/umad.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
   --> packages/ec-linear/src/mutator/umad.rs:128:13
    |
128 |             long_index += 1;
    |             ^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/ec-linear/src/mutator/umad.rs#L130C48-L130C87">packages/ec-linear/src/mutator/umad.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
   --> packages/ec-linear/src/mutator/umad.rs:130:48
    |
130 |         (short_index == short.len()).then_some(num_missing + (long.len() - long_index))
    |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/push/src/genome/plushy.rs#L64C36-L64C85">packages/push/src/genome/plushy.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
  --> packages/push/src/genome/plushy.rs:64:36
   |
64 |             1.0 / f32::conv_approx(instructions_distribution.num_choices().get() + 1),
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
   = note: requested on the command line with `-W clippy::arithmetic-side-effects`
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/push/src/instruction/float.rs#L56C66-L56C71">packages/push/src/instruction/float.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
  --> packages/push/src/instruction/float.rs:56:66
   |
56 |                         if y == 0.0 { OrderedFloat(1.0) } else { x / y }
   |                                                                  ^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/push/src/push_vm/stack.rs#L219C22-L219C37">packages/push/src/push_vm/stack.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
   --> packages/push/src/push_vm/stack.rs:219:22
    |
219 |                 .get(self.size() - 2)
    |                      ^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/push/src/push_vm/stack.rs#L365C12-L365C36">packages/push/src/push_vm/stack.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
   --> packages/push/src/push_vm/stack.rs:365:12
    |
365 |         if iter.len() + self.size() > self.max_stack_size {
    |            ^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/push/tests/float.rs#L87C31-L87C36">packages/push/tests/float.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
  --> packages/push/tests/float.rs:87:31
   |
87 |         let expected_result = x + y;
   |                               ^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
   = note: requested on the command line with `-W clippy::arithmetic-side-effects`
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/push/tests/float.rs#L101C31-L101C36">packages/push/tests/float.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
   --> packages/push/tests/float.rs:101:31
    |
101 |         let expected_result = x - y;
    |                               ^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/push/tests/float.rs#L115C31-L115C36">packages/push/tests/float.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
   --> packages/push/tests/float.rs:115:31
    |
115 |         let expected_result = x * y;
    |                               ^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/push/tests/float.rs#L129C72-L129C77">packages/push/tests/float.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
   --> packages/push/tests/float.rs:129:72
    |
129 |         let expected_result = if y == 0.0 { OrderedFloat(1.0) } else { x / y };
    |                                                                        ^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
```

</details>
<details>
<summary>
⚠️ arithmetic operation that can potentially result in unexpected side-effects
 -- <a href="https://github.com/JustusFluegel/rust-ga/blob/8576fd7da643fd30f1e64b33d51b29a2afb7b063/packages/push/tests/int.rs#L113C64-L113C66">packages/push/tests/int.rs</a>
</summary>

```txt
warning: arithmetic operation that can potentially result in unexpected side-effects
   --> packages/push/tests/int.rs:113:64
    |
113 |         prop_assert_eq!(*result.stack::<i64>().top().unwrap(), -x);
    |                                                                ^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
    = note: requested on the command line with `-W clippy::arithmetic-side-effects`
```

</details>

### PR DESCRIPTION
**Pull request by JustusFluegel**
This mostly deals with warnings in 2 ways:
- In library code either switches to saturating or checking operations
- In test code the lint is ignored.

Closes #160

### CODE REVIEW

File: packages/push/src/push_vm/stack.rs [hunk 1]
```diff
@@ -212,22 +212,19 @@ impl<T> Stack<T> {
...
-        } else {
-            Err(StackError::Underflow {
+        let index_second_to_top = self.size().checked_sub(2).ok_or(StackError::Underflow {
+            num_requested: 2,
+            num_present: self.size(),
+        })?;
+        let x = self.top()?;
+        let y = self
+            .values
+            .get(index_second_to_top)
```
Comment (NicMcPhee) [inline packages/push/src/push_vm/stack.rs:222]: We end up essentially checking the that there are at least two elements twice here, right? (Once here and once on line 215.) I doubt it matters a whole lot, but I wanted to make sure.

File: packages/push/src/push_vm/stack.rs [hunk 2]
```diff
@@ -362,7 +359,11 @@ impl<T> Stack<T> {
     {
         let iter = iter.into_iter();
         // Check that adding these items won't overflow the stack.
-        if iter.len() + self.size() > self.max_stack_size {
+        if iter
+            .len()
+            .checked_add(self.size())
+            .is_some_and(|x| x > self.max_stack_size)
```
Comment (NicMcPhee) [inline packages/push/src/push_vm/stack.rs:365]: This doesn't seem quite right to me. If `iter.len().checked_add(self.size())` returns `None`, then that means that adding the current stack size and the new items will be larger than `usize`, and it seems that that should fail in some way. Am I missing something?

**Review by NicMcPhee — COMMENTED** [review]
This look really nice, but I do have a could of questions about the changes to `stack.rs`. Thanks!

File: packages/push/src/push_vm/stack.rs (hunk 1 above)
Comment (JustusFluegel) [reply packages/push/src/push_vm/stack.rs:222]: yes, but we did the same thing before with the if check - this .get() here should now never be able to fail, just the same as it was before

File: packages/push/src/push_vm/stack.rs (hunk 2 above)
Comment (JustusFluegel) [reply packages/push/src/push_vm/stack.rs:365]: good catch! this should probably be .is_none_or instead

**Review by NicMcPhee — APPROVED** [review]
Thanks!

**Pull request merged**